Throttle search completions

cstyles opened the pull request:
I noticed that we make an API call for every typed character in the search bar. This PR aims to reduce the number of calls while still providing (relatively) live feedback.

Open questions:
1. Debouncing vs. throttling: I originally set out to add debouncing but then I found [this great article](https://css-tricks.com/debouncing-throttling-explained-examples/) and decided that I personally prefer throttling when it comes to search suggestions. Still, either would be an improvement over neither.
2. Wait interval: I chose 300ms mostly arbitrarily. It's ultimately a trade off between load and user experience.

natematykiewicz commented on app/javascript/controllers/app/search_controller.js:
```diff
@@ -1,4 +1,5 @@
 import { Controller } from "stimulus"
+import throttle from "lodash"
```
In order to only pull in this one function instead of the whole library, we should do `lodash/throttle` to enable tree shaking.

cstyles replied:
Fixed in 2ced93c11aa18a280d8179796fc2e264a49bbae4

natematykiewicz commented on app/javascript/controllers/app/search_controller.js:
```diff
@@ -74,6 +78,12 @@ export default class extends Controller {
       return
     }
 
+    if (this.lastQuery === query) {
+      return
+    }
+
+    this.lastQuery = query
```
What's this for?

cstyles replied:
Per the commit message from 2a4ec283c1a1f70971b271d5d485b3b866d2bec3:

> **Don't fetch new search results if the query wasn't modified**
> 
> This is useful since this method will still be invoked when, for
example, the user presses the arrow keys or selects the text with
Control+A. In these cases fetching more results is unnecessary so let's
exit early.

natematykiewicz replied:
That makes total sense to me. Thanks!

natematykiewicz commented on app/javascript/controllers/app/search_controller.js:
```diff
@@ -24,6 +25,9 @@ export default class extends Controller {
     </ul>
   `
+
+    this.throttledAutocomplete = _.throttle(this.autocomplete, 300)
```
Your recent changes look good, but you'll need to call throttle as the imported function `throttle`, instead of referencing it through `_`. Otherwise you get:

```
Error initializing controller
ReferenceError: Can't find variable: _
```

```suggestion
    this.throttledAutocomplete = throttle(this.autocomplete, 300)
```

cstyles replied:
Whoops, that's what I get for pushing changes without testing locally. Fixed in d09d7d515991476a9de095cfd0516335ab0d134c

colby-swandale commented on app/javascript/controllers/app/search_controller.js:
```diff
@@ -24,6 +25,9 @@ export default class extends Controller {
     </ul>
   `
+
+    this.throttledAutocomplete = _.throttle(this.autocomplete, 300)
```
To me, exposing the idea of throttling requests to the rest of the application is a leaky abstraction. Can we keep the details inside the implementation and keep the public API the same?

cstyles replied:
Good idea! Fixed in 6b85364cb12f5f0249a2798079010fae41f75d9a

cstyles commented:
Rebased onto master and resolved the merge conflict. No code changes.

<details>
<summary>range-diff</summary>

```
$ git range-diff origin/master 6b85364 c591e02
1:  598e1db ! 1:  0a98dd9 Add `lodash.throttle` package
    @@ Commit message
      ## package.json ##
     @@
          "@rails/webpacker": "^5.2.1",
    -     "clipboard-copy": "^3.1.0",
    -     "hotkeys-js": "^3.8.1",
    +     "clipboard-copy": "^4.0.1",
    +     "hotkeys-js": "^3.8.2",
     +    "lodash.throttle": "^4.1.1",
    -     "monaco-editor": "^0.21.2",
    -     "monaco-editor-webpack-plugin": "^2.0.0",
    -     "mustache": "^4.0.1",
    +     "monaco-editor": "^0.21.3",
    +     "monaco-editor-webpack-plugin": "^2.1.0",
    +     "mustache": "^4.1.0",

      ## yarn.lock ##
     @@ yarn.lock: lodash.templatesettings@^4.0.0:
2:  d8ca144 = 2:  8ea75f5 Throttle `autocomplete` calls to once every 300ms
3:  2a4ec28 = 3:  ed66d77 Don't fetch new search results if the query wasn't modified
4:  d09d7d5 = 4:  ad0148f Only pull in the `throttle` function instead of the entire library
5:  6b85364 = 5:  c591e02 Rename `throttledAutocomplete` back to `autocomplete`
```

</details>

colby-swandale commented:
Thanks for updating the PR. I'll try and get it reviewed in the next couple of days.

colby-swandale reviewed (approved):
Thanks for contributing this change and apologies for taking so long getting this merged! 🙇

colby-swandale commented:
@natematykiewicz are you able to review this again and resolve the Request Change if you're happy with this? Thanks!

cstyles commented:
Rebased, resolved the merge conflict again, and removed the merge commit (900f6c9c63f9709fb10e9806c4d65815d145dcc7). No code changes:

<details>
<summary>range-diff</summary>

```
$ git range-diff origin/master @{upstream} HEAD
1:  0a98dd9 ! 1:  d3aeed3 Add `lodash.throttle` package
    @@ package.json
     @@
          "@rails/webpacker": "^5.2.1",
          "clipboard-copy": "^4.0.1",
    -     "hotkeys-js": "^3.8.2",
    +     "hotkeys-js": "^3.8.3",
     +    "lodash.throttle": "^4.1.1",
          "monaco-editor": "^0.21.3",
          "monaco-editor-webpack-plugin": "^2.1.0",
2:  8ea75f5 = 2:  28667e4 Throttle `autocomplete` calls to once every 300ms
3:  ed66d77 = 3:  3963ff9 Don't fetch new search results if the query wasn't modified
4:  ad0148f = 4:  65a380e Only pull in the `throttle` function instead of the entire library
5:  c591e02 = 5:  599fdcd Rename `throttledAutocomplete` back to `autocomplete`
```

</details>

natematykiewicz reviewed (approved):
LGTM @colby-swandale
